### PR TITLE
Feature/ap 8

### PR DESCRIPTION
**Pull request by rdhawladar**

### CODE REVIEW

File: internal/app/model/entities.go [hunk 1]
```diff
@@ -1 +1,38 @@
...
+	DestinationAddress string  `json:"destination_address" db:"destination_address"`
+	SourceTime         string  `json:"source_time" db:"source_time"`
+	ParcelType         string  `json:"type" db:"type"`
+	Price              float32 `json:"price" db:"price"`
+	CarrierFee         float32 `json:"carrier_fee" db:"carrier_fee"`
+	CompanyFee         float32 `json:"company_fee" db:"company_fee"`
+}
+
+func (p *Parcel) ValidateParcelInput() error {
+
```
Comment (shuza) [inline internal/app/model/entities.go]: remove this blank line

File: internal/app/model/response.go [hunk 2]
```diff
@@ -7,6 +7,7 @@ import (
 
 var ErrNotFound = fmt.Errorf("not found")
 var ErrInvalid = fmt.Errorf("invalid")
+var ErrEmpty = fmt.Errorf("empty") // Empty error
```
Comment (shuza) [inline internal/app/model/response.go]: remove comment

File: internal/app/parcel/repository.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	"parcel-service/internal/app/model"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/lib/pq"
+)
+
+// SQL Query and error
+const (
+	errUniqueViolation = pq.ErrorCode("23505")
+	insertParcelQuery  = `INSERT INTO parcel (user_id, carrier_id, source_address, destination_address, source_time, type, price, carrier_fee, company_fee) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)`
```
Comment (shuza) [inline internal/app/parcel/repository.go]: at the time of parcel creation carrier_id will be empty so remove carrier_id from query

File: internal/app/parcel/repository.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+func NewRepository(db *sqlx.DB) *repository {
+	return &repository{
+		db: db,
+	}
+}
+
+func (r *repository) InsertParcel(ctx context.Context, parcel model.Parcel) error {
+	// actualPhone := RMCodeAndSpace(user.PhoneNumber)
+	fmt.Print(parcel.CarrierFee)
+	if _, err := r.db.ExecContext(ctx, insertParcelQuery, parcel.UserID, 1, parcel.SourceAddress, parcel.DestinationAddress, parcel.SourceTime, parcel.ParcelType, 200.0, 180.0, 20.0); err != nil {
```
Comment (shuza) [inline internal/app/parcel/repository.go]: don't hardcode price in repository. send it from service layer

File: internal/app/parcel/service.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+	"context"
+	"parcel-service/internal/app/model"
+	svc "parcel-service/internal/app/service"
+)
+
+type service struct {
+	repo svc.ParcelRepository
+}
+
+// NewService is to generate for new repo
```
Comment (shuza) [inline internal/app/parcel/service.go]: remoce comment

File: internal/app/parcel/service.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+func NewService(repo svc.ParcelRepository) *service {
+	return &service{
+		repo: repo,
+	}
+}
+
+// CreaetParcel is to hash password, validate credentials and inserting data into database
```
Comment (shuza) [inline internal/app/parcel/service.go]: remove comment

File: internal/app/parcel/service.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+func (s *service) CreateParcel(ctx context.Context, parcel model.Parcel) error {
+	/* if p, err := util.HashPassword(user.Password); err == nil {
```
Comment (shuza) [inline internal/app/parcel/service.go]: if don't needed remove those code instead of comment out

File: internal/app/server/halder_test.go [hunk 8]
```diff
@@ -1 +1,91 @@
...
+	"parcel-service/internal/app/service/mocks"
+	"strings"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/gorilla/mux"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
```
Comment (shuza) [inline internal/app/server/halder_test.go]: don't use const at here. its only used by one test case so put it inside that test case

File: internal/app/server/handler.go [hunk 9]
```diff
@@ -1 +1,36 @@
...
+
+func (s *server) createParcel(w http.ResponseWriter, r *http.Request) {
+	var data model.Parcel
+
+	if err := json.NewDecoder(r.Body).Decode(&data); err != nil {
+		ErrUnprocessableEntityResponse(w, "Decode Error", err)
+		return
+	}
+
+	// validating input credentials for Parcel create
```
Comment (shuza) [inline internal/app/server/handler.go]: remove comment. single line of code you be readable without comment

**Review by shuza — CHANGES_REQUESTED** [review]
refactor codes according to comments and remove comments for single line of code. single line of code should be readable and understandable without comment

File: internal/app/server/handler.go [hunk 10]
```diff
@@ -1 +1,44 @@
...
+	"errors"
+	"net/http"
+	"parcel-service/internal/app/model"
+
+	"github.com/rs/zerolog/log"
+)
+
+func (s *server) createParcel(w http.ResponseWriter, r *http.Request) {
+	const (
+		CARRIER_FEE = 180.00
```
Comment (shuza) [inline internal/app/server/handler.go]: this should be in service layer. later service will call pricing service (external service) to get those value

**Review by shuza — CHANGES_REQUESTED** [review]
move pricing constant in service layer. layer service layer will call external component to get pricing data

File: internal/app/server/handler.go [hunk 11]
```diff
@@ -1 +1,44 @@
...
+	data.CarrierFee = CARRIER_FEE
+	data.CompanyFee = COMPANY_FEE
+	data.Price = CARRIER_FEE + COMPANY_FEE
+
+	if err := s.parcelService.CreateParcel(r.Context(), data); err != nil {
+		if errors.Is(err, model.ErrInvalid) {
+			ErrInvalidEntityResponse(w, "invalid parcel", err)
+			return
+		}
+		log.Error().Err(err).Msgf("[signUp] failed to create parcel Error: %v", err)
```
Comment (zrshishir) [inline internal/app/server/handler.go]: at line no 39 `[signUp]` should be chaged to `[parcel]` route.

**Review by zrshishir — CHANGES_REQUESTED** [review]

File: internal/app/model/entities.go [hunk 12]
```diff
@@ -1 +1,37 @@
...
+import "fmt"
+
+type Parcel struct {
+	ID                 int     `json:"id"`
+	UserID             int     `json:"user_id" db:"user_id"`
+	CarrierID          int     `json:"carrier_id" db:"carrier_id"`
+	Status             int     `json:"status"`
+	SourceAddress      string  `json:"source_address" db:"source_address"`
+	DestinationAddress string  `json:"destination_address" db:"destination_address"`
+	SourceTime         string  `json:"source_time" db:"source_time"`
```
Comment (zrshishir) [inline internal/app/model/entities.go]: `source_time` should be timestamp not string.

**Review by zrshishir — CHANGES_REQUESTED** [review]

**Review by zrshishir — APPROVED** [review]

**Review by shuza — APPROVED** [review]
lgtm

**Pull request merged**